npep/npep-126-egress-traffic-control.md

```diff
@@ -0,0 +1,93 @@
+    - ANP currently supports only east-west traffic and this traffic flow cuts from
+    overlay to underlay which makes this part of the egress (northbound) use case.
+    - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+    - NOTE: Traffic towards nodes here includes traffic towards host-networked pods on that node
+      because a "node" resource encompasses all objects that share the host-networking resources
+* Implement egress traffic control towards hostNetworked pods
+  - Currently the behaviour for policies defined around traffic from cluster
+  workloads (non-hostNetworked pods) towards hostNetworked pods in the
+  cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+  - ANP currently supports only east-west traffic and this traffic flow cuts from
+  overlay to underlay which makes this part of the egress (northbound) use case.
+  - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+  - NOTE: We can keep this goal experimental in case implementors cannot distinguish between
+    a node and host-networked pod
+
+## Non-Goals
+
+* Implementing southbound (ingress) traffic use cases is outside the scope of this NPEP
+
+## Introduction
+
+### User Stories for egress traffic control towards external destinations
+
+1. **As a** cluster administrator **I want** to restrict traffic from
+specific cluster workloads to all or specific destinations outside the
+cluster **so that** I can enforce security for northbound traffic.
+Example: Pods in namespaceA and namespaceB should not be able to talk
+to the internet but they should be able to access company's intranet.
+
+2. **As a** cluster administrator **I want** to explicitly allow traffic
+from cluster workloads to handpicked destinations outside the cluster and
+deny the rest **so that** I can guarantee cluster wide egress traffic
+flow irrespective of whatever network policies are defined by the
+developer for specific namespaces. Example: As an owner of namespaceA I
+define policies that deny all northbound egress traffic for that
+namespace. However the cluster-admin can decide all namespaces in the
+cluster must be able to talk to the EXTERNAL_DNS_SERVER_IP on port 53.
+
+3. **As a** cluster administrator **I want** to explicitly delegate policy
+decisions for traffic from specific cluster workloads to all or specific
+destinations outside the cluster to the namespace owners or application
+developers **so that** the existing network policies that match these same set
+of destinations (a.k.a `ipBlock` in NetworkPolicyV1 world) can take effect
+Example: Egress traffic destined towards 8.8.8.8 from pods in namespaceA is
+delegated to the Kubernetes network policies implemented in namespaceA. If no
+Kubernetes network policies are implemented, the traffic will be allowed.
+(User story taken from the original ANP KEP for `Pass` action)
```

Comment:
Yeah, this isn't really a user story... It's too abstract.

The goal of user stories is not to find a set of tiles that exactly covers the proposed API. It's to show actual real-world scenarios where the proposed API will solve someone's problem (and where a different API, or no new API at all, would leave the problem unsolved).

If we don't have a real-world example that justifies the existence of `Pass` for cluster egress policies, then we don't have a user story for `Pass`. That doesn't mean we have to get rid of `Pass`, but it means we could be open to considering implementations that don't provide it. (That would be very unlikely just because it would be harder for us to _not_ provide `Pass` in cluster egress policies than it would be to just leave it in. But, eg, maybe it turns out there is some security problem with allowing `Pass` in cluster egress policies that doesn't apply to cluster-internal policies. And in that case, it would be good for us to know that we _can_ get rid of `Pass` if we need to.)

Reply:
good point! and I was talking to a customer last week where they pointed out, instead of using `Pass` I can just not have an ANP at all and the corresponding NP will just take effect anyways sort of like how NP/BANP naturally work without passing...

Reply:
Only reason I kept Pass was for compatibility with how east-west works, so let me think a bit more and fish around for the pass use case. I'll also bring this up in sig-net-pol u/s meeting today